### PR TITLE
RM-8726 AssemblyFinderTypeDiscoveryService.GetTypes performs very bad

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-8706

### CODE REVIEW

File: Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs [hunk 1]
```diff
@@ -69,13 +70,19 @@ public static BaseTypeCache Create (IEnumerable<Type> types)
             } while (nextBaseType != null);
           }
 
-          foreach (var interfaceType in type.GetInterfaces())
+          var implementedInterfaces = type.GetInterfaces();
+          if (implementedInterfaces.Length > 0)
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs:74]: Optimization vs code question:
Is there a benefit in guarding the entry? The clear-operation and the loop aren't really an issue for empty lists but we got more conditions so the code is more complex. thoughts?

File: Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs [hunk 2]
```diff
@@ -69,13 +70,19 @@ public static BaseTypeCache Create (IEnumerable<Type> types)
...
-              typesToProcess.Enqueue(normalizedInterfaceType);
+            processedInterfaces.Clear();
+            foreach (var implementedInterface in implementedInterfaces)
+            {
+              var normalizedInterfaceType = GetNormalizedType(implementedInterface);
+              if (enqueuedTypes.Add(normalizedInterfaceType))
+                typesToProcess.Enqueue(normalizedInterfaceType);
 
-            interfaces.Add(new KeyValuePair<Type, Type>(normalizedInterfaceType, type));
+              if (processedInterfaces.Add(normalizedInterfaceType))
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs:83]: What did you learn about the performance characteristicts for small Hashsets?

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
Got questions about simplifying things and Perf-results

File: Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs (hunk 1 above)
Comment (a-ctor) [reply Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs]: I checked and removed the if

File: Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs (hunk 2 above)
Comment (a-ctor) [reply Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs]: Too much variability to make a decisive decision so I choose the simpler one with the hash set.

File: Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs [hunk 3]
```diff
@@ -69,13 +70,15 @@ public static BaseTypeCache Create (IEnumerable<Type> types)
             } while (nextBaseType != null);
           }
 
-          foreach (var interfaceType in type.GetInterfaces())
+          processedInterfaces.Clear();
+          foreach (var implementedInterface in type.GetInterfaces())
           {
-            var normalizedInterfaceType = GetNormalizedType(interfaceType);
+            var normalizedInterfaceType = GetNormalizedType(implementedInterface);
```
Comment (MichaelKetting) [inline Remotion/Core/Core/Reflection/TypeDiscovery/BaseTypeCache.cs:76]: last question: change `implementedInterface` back to `interfaceType` to reduce change set?

**Review by MichaelKetting — APPROVED** [review]

**Review by MichaelKetting — CHANGES_REQUESTED** [review]
Forgot something: do we have a test for the multiple-generic interfaces situation?

**Comment by a-ctor** [comment]
> Forgot something: do we have a test for the multiple-generic interfaces situation?

We have a test that covers the scenario. That's why I introduced the `HashSet`'s in the first place :p

**Review by MichaelKetting — APPROVED** [review]

**Pull request merged**